common/models/templates/src/tokenization.ts

```diff
@@ -0,0 +1,91 @@
+// While we _could_ define this within @keymanapp/models-wordbreakers instead, it's probably
+// better to leave that package as _just_ the wordbreakers.  
+
+namespace models {
+  export interface Tokenization {
+    /**
+     * An array of tokens to the left of the caret.  If the caret is in the middle of a token,
+     * only the part to the left of the caret is included.
+     */
+    left: USVString[],
+
+    /**
+     * An array of tokens to the right of the caret.  If the caret is in the middle of a token,
+     * only the part to the right of the caret is included.
+     */
+    right: USVString[],
+    
+    /**
+     * A flag indicating whether or not the caret's position in the context caused a token
+     * to be split.  If `true`, the last entry of `left` is from the same token as the first
+     * entry of `right`.
+     */
+    caretSplitsToken: boolean
+  }
+
+  export function tokenize(wordBreaker: WordBreakingFunction, context: Context): Tokenization {
+    context = context || {
+      left: undefined,
+      startOfBuffer: undefined,
+      endOfBuffer: undefined
+    };
+
+    let leftSpans  = wordBreaker(context.left || '') || [];
+    let rightSpans = wordBreaker(context.right || '') || [];
+
+    let tokenization: Tokenization = {
+      left: leftSpans.map(span => span.text),
+      right: rightSpans.map(span => span.text),
+      
+      // A default initialization of the value.
+      caretSplitsToken: false
+    };
+
+    // Now the hard part - determining whether or not the caret caused a token split.
+    if(leftSpans.length > 0 && rightSpans.length > 0) {
+      let leftTail = leftSpans[leftSpans.length - 1];
+      let rightHead = rightSpans[0];
+
+      // If tokenization includes all characters on each side of the caret,
+      // we have a good candidate for a caret-splitting scenario.
+      let leftSuffixWordbreak = leftTail.end != context.left.length;
```

Comment:
> What should happen when `context.left === undefined`?

Impossible; this is within the following `if`-block, starting at L45:
```
if(leftSpans.length > 0 && rightSpans.length > 0) {
```

`leftSpans.length === 0` when `context.left === undefined`.  Though, I suppose that is a bit of an assertion - while it holds for our defined wordbreakers, it's possible someone could define a custom wordbreaker that invalides the assertion.

> Are the lengths in code units (as in JavaScript) or in code points (as in KMW)?

Code units.  It's the same as what your `Span` class already uses, since that's what the tokenization algorithm works upon.

> Why bother using this comparison when later on the code uses `wordBreaker(leftTail.text + rightHead.text).length === 1`?

Because said "later code" doesn't check for any intervening whitespace characters that may have been removed.

Suppose a hypothetical situation where we want to tokenize the following English sentence:  "He made that butter fly", with the caret after "butter"s r.  (Why someone would actually pitch a stick of butter, as if it were a baseball, is beyond me.)

`wordBreaker("butter"+"fly")` => `wordBreaker("butterfly")` will have `.length === 1` - even if we're using a dictionary-backed approach, as "butterfly" is, itself, a valid English word.  On the other hand...

```
      if(leftSuffixWordbreak || rightPrefixWordbreak) {
        // Bypass the final test, as we already know the caret didn't split a token.
        // (The tokenization process already removed characters between the two.)
        return tokenization;
      }
```

With left = "butter" and right = " fly", one of the two following scenarios should occur:
- `rightPrefixWordbreak === false` because `rightHead.start === 1`, not `===0`.
- `rightHead` represents `''`, an empty token preceding the `"fly"` token.  In this case, the empty string is appended to / merged with `leftTail`, which has no net effect.

In either case, this earlier check catches a scenario that the later check fails on.